### PR TITLE
Fix #2108: Defer identifier name mangling to the emitter.

### LINKED ISSUE #2108: Defer identifier name mangling until desugaring
Currently, the IR maintains a list of keywords that are reserved in JS, and forbids them in its identifiers. This is too low-level a concern for the IR, and is not robust with future versions of JS adding new reserved keywords. We also forbid leading digits.

We should defer the mangling currently performed by `mangleJSName` in [`JSEncoding`](https://github.com/scala-js/scala-js/blob/master/compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala) to the desugaring.

This a backward binary incompatible change, so it will be for the next major version.

### PR DESCRIPTION
**Pull request by sjrd**
We introduce specialized classes for Names in the IR, instead of using aliases of Strings. The new Names wrap an `EncodedName`, which is an `Array[Byte]` for the UTF-8 encoding of a name. They are optimized for frequent comparisons and hash-code access.

Simple names must be non-empty and can contain all Unicode code points, with the exception of `/ ; . [`.

Local names, label names and field names are just simple names. Class names are non-empty sequences of simple names separated by dots (`.`). The simple name of method names must not contain the code point `<` unless they are exactly `<init>` or `<clinit>`.

The above restrictions exactly match what the JVM allows, so our scalac back-end can directly translate all symbol names to the IR without further mangling, just like the JVM back-end.

`MethodName`s are not just the *simple* name of a method. They also encode the full signature, where param type refs and result type ref are kept as is. Therefore, `MethodName`s do not directly have an `EncodedName` representing them entirely.

---

The responsibility of mangling those arbitrary Unicode strings into JavaScript identifiers is delegated to the emitter, in particular in `JSGen`. The encoding is as follows:

* ASCII characters that are valid in JS identifiers are kept as is.
* The individual *bytes* of the UTF-8 encoding that are not valid ASCII characters are translated into `(b + 256).toChar`. It turns out that all code points in the range [256,512) are valid in JS identifiers (including as first char).
* For local names, a leading `$` is rewritten as `\u03b4` (greek small letter delta) so that it does not clash with "envFields".
* For class names, `.` is rewritten as `_` and `_` is rewritten as `\uff3f` (FULLWIDTH SPACING UNDERSCORE).
* For method names, param type refs and result type refs are encoded using the `I` or `Llong.Name` scheme, and separated with `__`, which is basically how they were encoded before.

As before, a number of common package names are compressed as short prefixes. The classes `java.lang.Object` and `java.lang.String` are compressed as `O` and `T`, respectively. This applies both to whole class names and within signatures.

---

A lot of code in this commit has been written with raw efficiency in mind. Names are manipulated very often, so it is essential that operations on them be as fast as possible.

### CODE REVIEW

**Comment by sjrd** [comment]
@gzm0 This is now ready for review.

**Comment by gzm0** [comment]
One higher level question about `MethodNames`: I do not understand what we gain from introducing yet another encoding of `(name, signature)` pair and exposing this encoding to clients of the IR.

It seems the JVM uses [another encoding for the signature][1] which doesn't seem to be easily portable to our IR's conception of `Name`.

So have you considered just making `MethodName` something like:

```
final case class MethodName(val simpleName: ???, val paramTypeRefs: List[TypeRef], resultTypeRef: Option[TypeRef])
```

[1]: https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.3

**Comment by gzm0** [comment]
Hmmm... So you are saying this is mainly a speed concern (I'm not sure I really buy the easier serialization argument: now that complexity is just in the name itself and serialization already knows how to serialize strings and type refs; to be fair, deduplication is a bit easier).

I'm a bit concerned that the optimization is premature / over optimized. I assume we do not have any performance data about this :-/

In my understanding when writing high performance code, we want to make sure we profile / measure so we only need to optimize the code that is relevant for the speed (and the rest can stay high-level / nice). So my concern here is that we are trying to optimize everything.

I would feel much more comfortable if the representation of a method name could remain internal to the `ir` package (I'm kind of OK to live with the dot separated nature of class names, because IIUC we never use it as semantic info). Do you think this would be achievable while retaining the suspected performance benefits? In this case the advantages and disadvantages would only be on the performance front and there I agree with your judgment that info retrieval is less important than comparison.

**Comment by sjrd** [comment]
Well, I quickly hacked together an additional commit where `MethodName`s truly are structured, and do not even have a smart representation for comparison purposes. And well ... my totally unscientific "benchmarks" suggest that the structured representation is strictly better on all the subtasks of the linker.

So that kind of defeats the performance argument of having an encoded name. I can clean it up and squash it if you like it better.

**Comment by gzm0** [comment]
Yes, I like it much more. But I would go even further in terms of simplifying:

- Consider case classes for method name kinds (instead of magic strings). Might be worse if these are directly translated from Scalac.
- Consider using a `String` instead of an `EncodedName`. I somehow doubt that we will see any performance improvement by using a byte array.

My main concern here is readability of the `Definitions` module: The verbosity of the code makes it very difficult to see what the code actually does. For example: I resorted to the description of this PR to see which characters are valid in which kind of name. Simply because finding the right `validateXname` method and parsing what it does is hard. This in turn is not good, because it removes "verifiable semantics" from the code: If a comment gets out of date, somebody might make wrong assumptions about what the code does. If the code is concise, a comment is likely not even necessary.

Long story short: I'd like to be sure that the verbose code is actually needed for performance.

Let me know what you think. I do understand that names are a very core part of linking and therefore might indeed have a relevant performance impact.

**Comment by sjrd** [comment]
> * Consider case classes for method name kinds (instead of magic strings). Might be worse if these are directly translated from Scalac.

`<init>` is directly translated from scalac, but that's not really a big deal.

The magic strings `<init>` and `<clinit>` are how the JVM works. I can turn them into case classes but a) it's going to re-create a gap with the JVM and b) I don't think it's going to simplify anything. Looking through the uses of `isConstructor` and `simpleName` in the linker, we never use both in the same place (which means we would never pattern-match on the case classes for name kinds) except in `JSGen.genName(MethodName)`. In addition, everywhere I *do* use `simpleName`, I want it to be the string `<init>` or `<clinit>` when that's what it is (for display names, etc.).

Considering all that, I don't think using case classes for `<init>` and `<clinit>` would be an improvement.

For reflective proxies, I am thinking about turning them into a `Boolean` flag in the `MethodName`, so that `resultTypeRef` wouldn't need to be an `Option`. It would be `VoidRef` for constructors (which is what it is on the JVM, and fits the expected type for its body) and `ClassRef(j.l.Object)` for reflective proxies (which has no equivalent on the JVM, but at least it fits the expected type of its body and the result type of calls to them). This would probably simplify a number of things. But I intend to submit that as a separate PR.

> * Consider using a String instead of an EncodedName. I somehow doubt that we will see any performance improvement by using a byte array.

There are several things to consider:

* Embedding a `String`, which itself contains an `Array[Char]`, inside a `Name` means that there are two levels of indirection to get to the contents.
* `String.equals()` does not leverage its `hashCode` as an early return, so we would still need to do that ourselves.
* Their native hash code has a very bad distribution, so we need to compute our own Murmur hash code ourselves anyway as well.
* The overwhelming majority of names (if not *all* names in a reasonable codebase) only contain ASCII characters. Using `String`s doubles the memory footprint of those names.
* Serializing and deserializing the strings encodes them as a bastard form of UTF-8 anyway. They might as well *be* UTF-8 in the first place.
* In the JVM, names are UTF-8 as well.

Now let's say we would use UTF-16 (though `String`s or not):

* We would still have to validate that the strings are correct UTF-16 (no unpaired surrogates). That's quantitatively easier than checking UTF-8, but not qualitatively easier.
* The `JSGen.genName` functions will become more complicated and less streamlined, because values above 0x00ff will need complicated treatment.
* and, well, that's about it; everything else would stay the same.

> My main concern here is readability of the Definitions module

I think that can be addressed quite easily, by:

* Adding clear doc-comments on the various `Name` subclasses, and
* Move the helper `validateXyz` methods *below* the definitions of the `Name` subclasses, so that the first thing we see is the meat, and not the details.

**Comment by gzm0** [comment]
OK. I think I'm convinced that the `SimpleMethodName` (with `<init>`, `<clinit>` as "special" `EncodedName`s) is the right approach here.

I'll review in detail. (feel free to squash the last four commits, I think there are also a couple of string matches remaining that are not necessary anymore).

**Comment by sjrd** [comment]
I'll do two commits out of the last four ones: one to introduce the convenience `MethodName` constructors, and one with everything else. That's mostly to reduce the load on the main (last) commit. I'll do that right away.

**Comment by sjrd** [comment]
> I'll do two commits out of the last four ones: one to introduce the convenience MethodName constructors, and one with everything else. That's mostly to reduce the load on the main (last) commit. I'll do that right away.

Well, that was not really helping, because there were too few occurrences and they were too intertwined with other changes of the main commit, so I squashed everything in one commit in the end.

**Comment by gzm0** [comment]
Please do not rebase this PR unless absolutely necessary. It destroys my review "viewed" history :)

**Comment by gzm0** [comment]
Ah, seems it *was* necessary. Fair enough :)

**Comment by sjrd** [comment]
Yes, it was necessary. There was a semantic conflict in GenJSCode after #3819.

I'm not touching it anymore, promise :p

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 1]
```diff
@@ -397,7 +400,7 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
       case ForIn(obj, keyVar @ LocalIdent(name, originalName), body) =>
         val newObj = transformExpr(obj)
         val newName = freshLocalName(name, mutable = false)
-        val newOriginalName = originalName.orElse(Some(name))
+        val newOriginalName = originalName.orElse(Some(name.nameString))
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: This is an example of a call site that handles `originalName` properly.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 2]
```diff
@@ -4402,21 +4408,28 @@ private[optimizer] object OptimizerCore {
    *  TODO It would be better for `Binding`s to represent the `this` reference
    *  using a separate mechanism.
    *
-   *  `this` is normally not a valid `LocalName`, because that is a JS keyword,
-   *  so we cast here.
+   *  We use an unsafely-constructed invalid `LocalName`, so that it never
+   *  clashes with normal local names. It is invalid because of the `;`.
+   */
+  private val LocalThisName =
+    LocalName.unsafeCreate("this;".map(_.toByte).toArray)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: I'm OK with this to unblock this PR, but I'd like to see it go away soon. Could you file a separate issue about this?

A concern I have is that a lot of code in `Encoding` makes the (reasonable) assumption that names are correct. In the current state, this wrong name will not break anything AFAICT but it might easily in the future. Better stick to the interface definitions :)

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 3]
```diff
@@ -5349,57 +5381,109 @@ private[optimizer] object OptimizerCore {
...
+          result = createNewName(baseEncoded, i)
         }
         usedNamesToNextCounter =
           usedNamesToNextCounter.updated(base, i + 1).updated(result, 1)
-        createNameFun(result)
+        result
+      }
+    }
+
+    private def createNewName(baseEncoded: Array[Byte], suffix: Int): N = {
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: Can we just use `withSuffix` of the relevant names here? Again, I doubt that removing the branch of non-ascii changes anything for a reasonable codebase due to branch prediction.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 4]
```diff
@@ -89,190 +452,280 @@ object Definitions {
...
-          case 'S' => ShortRef
-          case 'I' => IntRef
-          case 'J' => LongRef
-          case 'F' => FloatRef
-          case 'D' => DoubleRef
-          case 'N' => NullRef
-          case 'E' => NothingRef
-          case _   => ClassRef(ClassName(encodedName.substring(arrayDepth)))
-        }
+  private def encodeStringNameWithNonASCII(name: String): EncodedName = {
```
[the remark this answers is not in the document]
Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Just `String.getBytes(UTF_8)` is not correct here, because it does not report invalid inputs (unpaired surrogates). We have to use a `UTF_8.newEncoder()` manually and catch the `CharacterCodingException` that it throws, which is a bit of a pain. I've updated the code to use that, but I'm not sure the result is better. I can revert to the old version if you prefer.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 5]
```diff
@@ -89,190 +452,327 @@ object Definitions {
...
+
+  private def decodeStringName(encoded: EncodedName): String =
+    decodeStringName(encoded, 0, encoded.length)
+
+  private def decodeStringName(encoded: EncodedName, start: Int, end: Int): String = {
+    // scalastyle:off return
+    /* We optimistically assume that all characters are single-byte (i.e., in
+     * the ASCII repertoire), and fall back to a full UTF-8 decoder if we find
+     * a multi-byte character.
+     */
```
[the remark this answers is not in the document]
Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: It's used by `nameString`, which, as you have noticed, is used more than we'd like, due to original names.

Anyway, I think we should keep the optimistic path if only by symmetry with `encodeStringName`.

File: ir/src/test/scala/org/scalajs/ir/PrintersTest.scala [hunk 6]
```diff
@@ -593,22 +613,22 @@ class PrintersTest {
...
+    assertPrintEquals("new (f().test.Test::C)(4, 5)",
+        JSNew(JSPrivateSelect(fApplied, "test.Test", "C"), List(i(4), i(5))))
     assertPrintEquals("""new (f()["C"])()""",
         JSNew(JSSelect(fApplied, StringLiteral("C")), Nil))
   }
 
   @Test def printJSPrivateSelect(): Unit = {
-    assertPrintEquals("x.Ltest_Test::f",
-        JSPrivateSelect(ref("x", AnyType), "Ltest_Test", "f"))
+    assertPrintEquals("x.test.Test::f",
```
[the remark this answers is not in the document]
Comment (sjrd) [reply ir/src/test/scala/org/scalajs/ir/PrintersTest.scala]: The printed form of the IR is full of ambiguities. If we want to make it non-ambiguous, that require sweeping changes.

File: javalanglib/src/main/scala/java/lang/StackTrace.scala [hunk 7]
```diff
@@ -188,7 +188,8 @@ private[lang] object StackTrace {
    */
   private def extractClassMethod(functionName: String): js.Array[String] = {
     val PatC = """^(?:Object\.|\[object Object\]\.)?\$c_([^\.]+)(?:\.prototype)?\.([^\.]+)$""".re
-    val PatS = """^(?:Object\.|\[object Object\]\.)?\$(?:ct|ps?|s|f)_((?:_[^_]|[^_])+)__([^\.]+)$""".re
+    val PatS = """^(?:Object\.|\[object Object\]\.)?\$(?:ps?|s|f)_((?:_[^_]|[^_])+)__([^\.]+)$""".re
+    val PatCT = """^(?:Object\.|\[object Object\]\.)?\$ct_((?:_[^_]|[^_])+)__([^\.]*)$""".re
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalanglib/src/main/scala/java/lang/StackTrace.scala]: The encoding of `init` methods slightly changes in this PR, because the `init___` part of it disappeared (but we gained a trailing `__` in exchange), so `$ct_xyz` functions do not have the same format as `$p_`, `$s_` etc. functions. The actual difference in the regexp is the very last `*` instead of a `+`.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala [hunk 8]
```diff
@@ -106,15 +105,13 @@ object Analysis {
...
       }
 
-      val (simpleName, paramTypes, resultType) =
-        ir.Definitions.decodeMethodName(encodedName)
-
-      simpleName + "(" + paramTypes.map(typeRefDisplayName).mkString(",") + ")" +
-      resultType.fold("")(typeRefDisplayName)
+      encodedName.simpleName.nameString + "(" +
+      encodedName.paramTypeRefs.map(typeRefDisplayName).mkString(",") + ")" +
+      encodedName.resultTypeRef.fold("")(typeRefDisplayName)
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala]: The representation here is for human readers. The changes in the code in this PR preserve the output of `displayName` between before and after this commit.

The question should be: do we a different representation *in `MethodName.nameString` than this one? That's arguable, but for now I would say yes because `displayName` is ambiguous (in theory, there could exist a *class* named `int` on the classpath) whereas `nameString` is non-ambiguous.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala [hunk 9]
```diff
@@ -247,9 +248,9 @@ object Infos {
         case ArrayType(_) =>
           /* The pseudo Array class is not reified in our analyzer/analysis,
            * so we need to cheat here.
-           * In the Array[T] class family, only clone__O is defined and
-           * overrides j.l.Object.clone__O. Since this method is implemented
-           * in CoreJSLib and always kept, we can ignore it.
+           * In the Array[T] class family, only clone()Object is defined and
+           * overrides j.l.Object.clone()Object. Since this method is
+           * implemented in CoreJSLib and always kept, we can ignore it.
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala]: The `clone__O` encoding is now an implementation detail of the Emitter. It doesn't make sense for `Infos` to talk about methods using that encoding. So yes, IMO the comment needs to be updated in this commit.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala [hunk 10]
```diff
@@ -456,9 +457,9 @@ object Infos {
     private val builder = new ReachabilityInfoBuilder
 
     def generateMethodInfo(methodDef: MethodDef): MethodInfo = {
-      val (_, params, result) = decodeMethodName(methodDef.encodedName)
-      params.foreach(builder.maybeAddReferencedClass)
-      result.foreach(builder.maybeAddReferencedClass)
+      val encodedName = methodDef.encodedName
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala]: Yes, I think we'll need a pass at all the existing `encodedName`s to give them better names, but I'd keep that for a separate PR. I updated the local `val` here anyway, because at least we shouldn't *introduce* more `encodedName`s in this commit.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala [hunk 11]
```diff
@@ -397,8 +397,9 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
       if flags.namespace.isStatic
     } yield {
       implicit val pos = field.pos
-      val fullName = tree.encodedName + "__" + name
-      envFieldDef("t", fullName, genZeroOf(ftpe), origName, flags.isMutable)
+      val fullName = genName(tree.encodedName) + "__" + genName(name)
+      envFieldDef("t", fullName, genZeroOf(ftpe), origName, flags.isMutable,
+          keepFunctionExpression = false)
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala]: Yes, it's the best I could do :-/

File: linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/Trees.scala [hunk 12]
```diff
@@ -43,13 +43,39 @@ object Trees {
...
+      var i = 1
+      while (i != len) {
+        val c = name.charAt(i)
+        if (c != '$' && !Character.isUnicodeIdentifierPart(c))
+          return false
+        i += 1
+      }
+      true
+      // scalastyle:on return
+    }
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/Trees.scala]: No, it is still used by a few things, notably `JSGlobalRef`. But there is a followup PR about that at #3823.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 13]
```diff
@@ -1435,11 +1438,11 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
             Throw(New(ClassRef(NullPointerExceptionClass),
                 MethodIdent(NoArgConstructorName), Nil))).toPreTransform)
       case _ =>
-        if (isReflProxyName(methodName)) {
+        if (methodName.isReflectiveProxy) {
           // Never inline reflective proxies
           treeNotInlined
         } else if (methodName == ObjectCloneName && canBeArray(treceiver)) {
-          // #3778 Never inline the `clone__O()` method if the receiver can be an array
+          // #3778 Never inline the `clone()j.l.Object` method if the receiver can be an array
```
[the remark this answers is not in the document]
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: I don't understand. Isn't this change precisely "updating the method name in the comment"?

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala (hunk 2 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: Filed as #3828.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 14]
```diff
@@ -14,65 +14,428 @@ package org.scalajs.ir
...
+      extends Name(encoded) with Comparable[LocalName] {
+
+    type ThisName = LocalName
+
+    override def equals(that: Any): Boolean = {
+      (this eq that.asInstanceOf[AnyRef]) || (that match {
+        case that: LocalName => equalsName(that)
+        case _               => false
+      })
+    }
```
[the remark this answers is not in the document]
Comment (gzm0) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Hahaha OK :) So let's leave it.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 15]
```diff
@@ -14,65 +14,428 @@ package org.scalajs.ir
...
+
+    /** Returns `true` iff this is the name of an instance constructor. */
+    def isConstructor: Boolean = kind == ConstructorKind
+
+    /** Returns `true` iff this is the name of a static initializer. */
+    def isStaticInitializer: Boolean = kind == StaticInitializerKind
+
+    /** Returns `true` iff this is the name of a reflective proxy. */
+    def isReflectiveProxy: Boolean = kind == ReflectiveProxyKind
```
[the remark this answers is not in the document]
Comment (gzm0) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Ok. So let's revisit this if `isReflectiveProxy` becomes a field (are you still considering this)?

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 16]
```diff
@@ -89,190 +451,234 @@ object Definitions {
...
+        i += 1
+      validateSimpleEncodedName(encoded, start, i, openAngleBracketOK = true)
+      i += 1 // goes to `len + 1` iff we successfully parsed the last segment
+    }
 
-  /** Encodes a method name from its full signature. */
-  def encodeMethodName(baseName: String, paramTypes: List[TypeRef],
-      resultType: Option[TypeRef]): MethodName = {
+    if (i == len) // the last segment was empty (including when `len == 0`)
+      throwInvalidEncodedName(encoded)
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/ir/Definitions.scala]: This only triggers when `len == 0` right? Otherwise the last call to `validateSimpleEncodedName` would fail. Use a `do ... while`?

```scala
var len = encoded.length
var i = 0
do {
  val start = i
  while (i != len && encoded(i) != '.')
    i += 1
  validateSimpleEncodedName(...)
  i += 1
} while (i < len)
```

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala [hunk 17]
```diff
@@ -89,190 +452,327 @@ object Definitions {
...
+      }
+    }
+    i
+    // scalastyle:on return
+  }
 
-  /** Encodes a method name from its full signature. */
-  def encodeMethodName(baseName: String, paramTypes: List[TypeRef],
-      resultType: Option[TypeRef]): MethodName = {
+  private def validateMultibyteCodePointAndGetByteLen(encoded: EncodedName,
```
[the remark this answers is not in the document]
Comment (gzm0) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Hmm... OK. Not super happy about that TBH (especially all this stuff is untested), but not part of the public interface, so OK for now I guess.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (gzm0) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: I like this much more, as it doesn't introduce encoding logic. Just boring code. I like boring code.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala]: Yep, fair enough.

File: linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala]: I meant the format of encoded name v.s. the format of display name. But I think this is OK (I meant the same below).

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala [hunk 18]
```diff
@@ -1995,16 +1990,16 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
             js.Apply(newReceiver DOT transformMethodIdent(method), newArgs)
 
           def genDispatchApply(): js.Tree =
-            genCallHelper("dp_" + methodName, newReceiver :: newArgs: _*)
+            genCallHelper("dp_" + genName(methodName), newReceiver :: newArgs: _*)
 
           def genHijackedMethodApply(className: ClassName): js.Tree = {
-            val fullName = className + "__" + methodName
+            val fullName = genName(className) + "__" + genName(methodName)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala]: No helper in `JSGen` for this? (original comment seems unaddressed)

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: I'm still considering it, but I would leave it for another PR. In this PR I have not fundamentally changed the "typing discipline" for constructors and reflective proxies (just names). Using a mandatory `resultTypeRef` for constructors and reflective proxies, while using a flag for the latter, would be a change of typing discipline which is not directly related to this PR.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala (hunk 16 above)
Comment (sjrd) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: No it also happens if the last byte of `encoded` is a `'.'`, which is why the comment says "when the last segment is empty". I have expanded the comment to make that clearer.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala (hunk 18 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala]: Oops, I forgot that comment, indeed. I have now added the appropriate helpers (two overloads of `envField` in `JSGen`, and an additional overload of `envFieldDef` in `ClassEmitter`).

**Comment by sjrd** [comment]
Updated.

File: ir/src/main/scala/org/scalajs/ir/Definitions.scala (hunk 16 above)
Comment (gzm0) [reply ir/src/main/scala/org/scalajs/ir/Definitions.scala]: Ah, right. All this manual indexing is hard to parse :P
